### PR TITLE
added Fidelity PDF import of trade confirmations

### PR DESCRIPTION
**Pull request by ggrossbe**
added Fidelity PDF import of trade confirmations (not monthly/quarterly statements)

### CODE REVIEW

**Comment by ggrossbe** [comment]
P.S. tests are running fine. I cannot run PP as program from Eclipse on my Mac. Buttons are not showing, ...

**Comment by buchen** [comment]
> P.S. tests are running fine. I cannot run PP as program from Eclipse on my Mac. Buttons are not showing, ...

Do you use the https://marketplace.eclipse.org/content/launch-configuration-dsl to generate the launch configuration? (I am developing on macOS so it does work there from within Eclipse)

**Comment by buchen** [comment]
@Nirus2000

**Comment by Nirus2000** [comment]
Hello @ggrossbe 
Thank you for your pull request.
That looks very promising.

First of all, however, the importer must be called in PP.
Please add the respective importers to the [PDFImportAssistant.java](https://github.com/portfolio-performance/portfolio/blob/23a649217b18dc614f07e34d4dc416c49ebd1083/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/PDFImportAssistant.java#L31) 👍🏻 

To keep the source code clean, we would welcome the removal of unused source code (commented out).
The [Contributing rules](https://github.com/portfolio-performance/portfolio/blob/master/CONTRIBUTING.md#pdf-importer) will help you how to structure this correctly, but it already looks very good. Good work. 👍🏻 
Have a look there and read them... maybe this will help you to structure the source better.

One question remains.
Why two different importers? (FidelityPDFExtractor and FidelityStockPlanPDFExtractor)

Please merge the two.

Regards
Alex

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                         * 12-18-23 02079K107 23348-I7W2H DESCRIPTION and
+                         * DISCLOSURES You Bought ALPHABET INC CAP STK CL C
+                         * Principal Amount 3,990.00 30 WE HAVE ACTED AS AGENT.
+                         * Settlement Amount 3,990.00 at 133.0000 Symbol: GOOG
+                         */
+
+                        .section("month", "day", "year") //
+                        .match("^\\d+-\\w+.* (?<month>\\d+)-(?<day>\\d+)-(?<year>\\d+) \\d+-\\d+-\\d+.*$")
+                        .assign((t, v) -> //
+                        t.setDate(asDate(v.get("day") + "." + v.get("month") + ".20" + v.get("year"))))
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java:57]: For example with contributing rules

```suggestion
                        // @formatter:off
                        // 20123-1XXXXX 1* WK# 01-04-21 01-06-21 46428Q109 20123-XXXXX
                        // @formatter:on
                        .section("date") //
                        .match("^.* WK# (?<date>[\\d]{2}\\-[\\d]{2}\\-[\\d]{2}) .*$")
                        .assign((t, v) -> t.setDate(asDate(v.get("date"))))
```

**Review by Nirus2000 — CHANGES_REQUESTED** [review]

**Comment by ggrossbe** [comment]
> One question remains.
> Why two different importers? (FidelityPDFExtractor and FidelityStockPlanPDFExtractor)

Merged now. There are two different document formats from different brokerage accounts. One is a regular account and one is a company stock plan account.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                        .section("date") //
+                        .match("^.* \\w{2}# (?<date>[\\d]{2}\\-[\\d]{2}\\-[\\d]{2}) .*$")
+                        .assign((t, v) -> t.setDate(asDate(v.get("date"))))
+
+                        .section("type", "gross", "tickerSymbol", "name") //
+                        .match("^You (?<type>Bought|Sold) (?<name>.*) Principal Amount\\s+(?<gross>[\\.,\\d]+)$") //
+                        .match("^(?!REFERENCE)(?!DESCRIPTION)(?<tickerSymbol>[A-Z]{2,5}).*$") //
+                        .assign((t, v) -> {
+                            t.setCurrencyCode(asCurrencyCode("USD"));
+                            v.put("currency", "USD");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java:67]: You can use the "$" to identify the currency. You can find the appropriate regex -> [here](https://github.com/portfolio-performance/portfolio/blob/master/CONTRIBUTING.md#regular-expressions) <-.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+                        .section("date") //
+                        .match("^Sale Date: (?<date>[\\w]{3}\\/[\\d]{2}\\/[\\d]{4})$")
+                        .assign((t, v) -> {
+                            String date = v.get("date");
+                            // convert month to camel case
+                            if (Character.isLetter(date.charAt(0)))
+                            {
+                                date = date.charAt(0) + date.substring(1, 3).toLowerCase() + date.substring(3);
+                            }
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java:135]: I think if you use the ExtractorUtils correctly, we could save this part.
What did you think?

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        int lastDot = value.lastIndexOf(".");
+        int lastComma = value.lastIndexOf(",");
+
+        // returns the greater of two int values
+        if (Math.max(lastDot, lastComma) == lastDot)
+        {
+            language = "en";
+            country = "US";
+        }
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java:205]: if all "asShares" are in en_US you can remove this and use
.assign((t, v) -> t.setShares(asShares(v.get("shares"), "en", "US"))),

The same for "asAmount"... 👍🏻 ...

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java (hunk 2 above)
Comment (ggrossbe) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java:67]: there is no $ sign in the text :(

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java (hunk 3 above)
Comment (ggrossbe) [reply name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/FidelityPDFExtractor.java]: no, as the DateFormatter expects "Dec" and the text from the bank statement is "DEC"

**Pull request closed without merging**